Update receiverhelper for requests that failed to be received

gizas opened the pull request:
#### Description

This is an ongoing work to add :
- In the endOp function of [ObsReport](receiver/receiverhelper/obsreport.go), errors are classified as either:
   - Downstream errors (refused) - when internal.IsDownstreamError(err) returns true
   - Internal errors - when the error is not a downstream error
- New metrics metricsotelcol_receiver_internal_errors_spans, otelcol_receiver_internal_errors_metric_points and otelcol_receiver_internal_errors_log_records  in case of inetrnal errors returned above
- Support in  OTLP receiver to handle  IsClientDisconnectError. Network disconnect errors will be reported in otelcol_receiver_internal_errors_metric_points 
- Support of otelcol_receiver_requests metric that will determines the outcome of the request:
    -  "success" if there's no error
    -   "refused" if there's a downstream error
    -  "failure" if there's an internal error

#### Link to tracking issue
Fixes # https://github.com/open-telemetry/opentelemetry-collector/issues/12207

#### Testing

Only added the integration tests. That is why opened this is as a draft

#### Documentation

See updated [receiver/receiverhelper/documentation.md](receiver/receiverhelper/documentation.md)

jade-guiton-dd commented:
@gizas Are you still working on this? Or should it be marked as ready for review?

gizas commented:
@jade-guiton-dd I will try to fix the linting errors and will open it as ready for review. thanks for the ping

jade-guiton-dd commented:
@gizas It looks like there are still linting errors and errors in go-generate, could you take a look at fixing them? (The contrib-tests failure seem to be because of tests relying on specific metric descriptions, so I think this can be ignored)

I took a first look at the PR, and I was thinking it's probably a good idea to wait until pipeline instrumentation (#12676) has implemented the "outcome = refused" case, so we can reuse the logic and types in this PR as well.

gizas commented:
@jade-guiton-dd neglect my above comment. I removed the otelcol_receiver_requests metric for now. 
I only kept the otelcol_receiver_internal_errors* metrics. 

I have added a new test case  [TestOTLPReceiverGRPCMetricsIngestTest](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/files#diff-3c2dc2ebf8936823c033ebd03c8196f2b2192962dcd70d165dbaab41bfd79d27R529) which I hope clarifies a little more the functionality

In case  pipeline instrumentation (https://github.com/open-telemetry/opentelemetry-collector/issues/12676), shall we implement the otelcol_receiver_requests metric?

jade-guiton-dd commented:
Sorry, I'm not sure I understand your question. I think your understanding is correct, but let me try to give a bit more context for issue #12207, since I wrote it a while ago and some things have changed since then. (Sorry in advance for the really long message.)

We've recently added support for pipeline instrumentation in the Collector, which adds an instrumentation layer between every consumer (processor/connector/exporter) and the component before it, in order to record the number of telemetry items passing through the `Consume` calls, as well as whether the call returned an error or not.

Ideally, to be able to identify where errors are coming from in a pipeline, we want a 3-way distinction; either:
- the call was a `success`
- the call returned an error, which signals a `failure` in the next consumer
- the next consumer processed the data successfully, but its output was `refused` by a component further downstream, and the call returned that downsteam error.

The distinction between these last two cases isn't yet implemented, but the idea will be to wrap errors in a "downstream" wrapper before returning them from the `Consume` instrumentation layer, which will signal to other instrumentation layers upstream that this error was already associated with a particular component.

One issue with pipeline instrumentation is that it doesn't allow users to monitor errors that occur inside receivers, before the first `Consume` call.

`receiverhelper.ObsReport` provides `StartXOp` and `EndXOp` methods, which currently output a `otelcol_receiver_refused_x` metric when an error is passed to the latter. There are multiple issues that prevent properly tracking internal errors with this mechanism:
1. When passing an error to `EndXOp`, it doesn't distinguish internal errors from downstream errors.
2. Comparing with the terminology used in pipeline instrumentation, the naming of the metric ("refused") implies that it only tracks downstream errors.
3. The `EndXOp` requires passing a number of telemetry items. This means that component developers either:
    - Wait until they've successfully received and parsed telemetry items before calling `StartXOp`, since only at that point are they able to count the number of items involved. This often means only wrapping `StartXOp`/`EndXOp` around the final `Consume` call, which means no internal errors are recorded. (This is what the OTLP receiver does.)
    - Call `StartXOp` early, but never call `EndXOp` in case of an internal error, which means no metric is recorded, and a span is never emitted. I would consider this a faulty use of the API. (This seems to be what the [carbonreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/faca52ae75164cf1ad75db6365b1c750fc36d5a2/receiver/carbonreceiver/reporter.go#L48) does.)
    - Call `StartXOp` early, but call `EndXOp` with an item count of zero in case of an internal error, which means a span is emitted, but no change to any metric is recorded. (This seems to be what the [otlpjsonfilereceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/faca52ae75164cf1ad75db6365b1c750fc36d5a2/receiver/otlpjsonfilereceiver/file.go#L92) does.)

I currently think the best way to solve these issues is the following:
- Standardize on the third way to use `ObsReport`: "call `StartXOp` as early as possible, and call `EndXOp` with an item count of zero if an internal error occured and no item count is known."
    We need to make sure this is well documented, and we will need to modify existing receivers to follow this guideline (in separate PRs of course).
- Inside `EndXOp`, check whether the error we receive has been tagged as "downstream" by pipeline instrumentation. Only emit `otelcol_receiver_refused_x` if yes.
- Add a generic `otelcol_receiver_requests` metric, which we increase by 1 for every `EndXOp` call, with an `outcome` attribute set to:
    - "success" if no error was passed in
    - "refused" if the error is tagged as coming from downstream (ie. the case where we emit `otelcol_receiver_refused_x`)
    - "failure" otherwise
- We can maybe also emit a `otelcol_receiver_internal_errors_x` metric like you did in this PR (although I think `otelcol_receiver_failed_x` would be more consistent). This will only be useful in cases where we have an internal error, but still got far enough to be able to count items. I think that's a somewhat niche case, but I guess it's better than throwing away the itemCount information.

jade-guiton-dd commented:
Unrelatedly, do you think it would be possible to split the changes in the receiver documentation / otlpreceiver from the internal telemetry changes in receiverhelper into separate PRs? I feel relatively qualified in reviewing the latter since I opened the issue, but not so much for the former.

jade-guiton-dd commented:
@gizas Sorry for the ping, but do you think you will keep working on this?

gizas commented:
Yes and no problem at all thanks for the ping. I have this work in my todo list but can not commit for a deadline for the next period. And thank you for the comments above, still in review!

jade-guiton-dd commented on .chloggen/custom_receiverhelper.yaml:
```diff
@@ -0,0 +1,25 @@
+component: receiverhelper
+
+# A brief description of the change.  Surround your text with quotes ("") if it needs to start with a backtick (`).
+note: Update receiverhelper StartXOp and EndXOp methods with otelcol_receiver_refused_x with otelcol_receiver_failed_x in order to be able to distinguish internal vs downstream errors
```
Typo:
```suggestion
note: Update receiverhelper StartXOp and EndXOp methods to emit otelcol_receiver_refused_x and otelcol_receiver_failed_x in order to be able to distinguish internal vs downstream errors
```

jade-guiton-dd commented on .chloggen/custom_receiverhelper.yaml:
```diff
@@ -0,0 +1,25 @@
+note: Update receiverhelper StartXOp and EndXOp methods with otelcol_receiver_refused_x with otelcol_receiver_failed_x in order to be able to distinguish internal vs downstream errors
+
+# One or more tracking issues or pull requests related to the change
+issues: ["https://github.com/open-telemetry/opentelemetry-collector/issues/12207"]
```
This should just be the issue number: (sorry, it's not obvious from the comment)
```suggestion
issues: [12207]
```

gizas replied:
Done

jade-guiton-dd commented on receiver/otlpreceiver/internal/errors/errors.go:
```diff
@@ -24,6 +40,12 @@ func GetStatusFromError(err error) error {
+	// and should lead to a retryable `codes.Unavailable`.
+	if IsClientDisconnectError(err) {
+		return status.New(codes.Unavailable, http.StatusText(http.StatusServiceUnavailable)).Err()
+	}
```
As I previously mentioned, I think it would be good to split all these changes to error handling in the OTLP receiver into a separate PR. As I understand it they are completely unrelated to #12207 and would benefit from a separate review.

gizas replied:
Rmoved in [fixing comments](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/d347bd199ab9767991169b999e47414724a4223c)

jade-guiton-dd commented on receiver/otlpreceiver/internal/errors/errors.go:
```diff
@@ -12,8 +12,24 @@ import (
 func GetStatusFromError(err error) error {
+	if err == nil {
+		return nil
+	}
```
Nit: I think this is unnecessary. According to the docs, `status.FromError` and `Status.Err` handle the `nil` case fine.

jade-guiton-dd commented on receiver/otlpreceiver/internal/errors/errors.go:
```diff
@@ -12,8 +12,24 @@ import (
 	"go.opentelemetry.io/collector/consumer/consumererror"
 )
 
+// IsClientDisconnectError returns true if the error indicates a client disconnection
```
Nit: Maybe clarify that we're not talking about the OTLP receiver's client, but a client connection later down the pipeline.

jade-guiton-dd commented on receiver/otlpreceiver/otlp_test.go:
```diff
@@ -1263,7 +1304,28 @@ func (esc *errOrSinkConsumer) checkData(t *testing.T, data any, dataLen int) {
 }
 
 func assertReceiverTraces(t *testing.T, tt *componenttest.Telemetry, id component.ID, transport string, accepted, refused int64) {
```
Nit: I think it would make sense to rename the `refused` parameter to `failed` since you changed its semantics

gizas replied:
Done

jade-guiton-dd reviewed (changes requested):
Thank you for all the work! I think it would be good to split this PR in two, and to use the downstream wrapper from #13234 once it gets merged. I also have some nitpicks regarding documentation.

gizas commented on receiver/receiverhelper/metadata.yaml:
```diff
@@ -68,3 +75,31 @@ telemetry:
+    type: int
+  failure:
+    description: Failed receiver requests
+    type: int
```
Sorry forgot about it. Changed

gizas commented on receiver/receiverhelper/obsreport.go:
```diff
@@ -160,34 +160,54 @@ func (rec *ObsReport) endOp(
 	// end span according to errors
 	if span.IsRecording() {
-		var acceptedItemsKey, refusedItemsKey string
+		var acceptedItemsKey, refusedItemsKey, internalErrorsKey string
```
[changing internalErrorsKey to failedItemsKey](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/36acb843502846008b382cd013bbe4b9fd19decd)

gizas commented on receiver/receiverhelper/internal/errors.go:
```diff
@@ -0,0 +1,19 @@
+// WrapDownstreamError wraps an error to indicate it's a downstream error
+func WrapDownstreamError(err error) error {
+	return errors.Join(ErrDownstreamError, err)
+}
```
Thank you will have a look for this

gizas commented on receiver/receiverhelper/internal/errors.go:
```diff
@@ -0,0 +1,19 @@
+
+// WrapDownstreamError wraps an error to indicate it's a downstream error
+func WrapDownstreamError(err error) error {
+	return errors.Join(ErrDownstreamError, err)
```
Will update this in combination to https://github.com/open-telemetry/opentelemetry-collector/pull/12802#discussion_r2222143721

gizas commented on receiver/receiverhelper/metadata.yaml:
```diff
@@ -85,21 +85,17 @@ telemetry:
-  success:
+  outcome:
     description: Successful receiver requests
     type: int
```
Fixed in [fix string in metadata and comment](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/535d2ba11a2ee898867cca79e1eb4dc3f383784a) thank you

jade-guiton-dd commented:
For the failure in `TestOCReceiverTrace_HandleNextConsumerResponse`, yes, I think that would be the right approach. May be more future-proof to add the downstream wrapper [in the `errOrSinkConsumer` implementation](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/8042df9258d0eeddfd145febe644cfe89817a4fa/receiver/opencensusreceiver/opencensus_test.go#L908) directly.

For the failures in `awsxrayreceiver/internal/udppoller/poller_test.go`, it looks like the errors it's testing for are internal errors, so we'll have to update [the `assertReceiverTraces` function](https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/8042df9258d0eeddfd145febe644cfe89817a4fa/receiver/awsxrayreceiver/internal/udppoller/poller_test.go#L508) to expect `otelcol_receiver_failed_spans` instead.

I was wondering if we should consider this PR a breaking change, since we're changing the semantics of an existing metric in a way which may break tests and dashboard (ie. errors that were previously surfaced under one metric are now under a different metric). I think the fact that we have 2 unit tests relying on the current behavior shows that we might want to be cautious with this. Maybe this change should be under a feature gate?

jade-guiton-dd commented:
> I was thinking the same on this that is why asked you. I agree that this is a breaking change

I also think this is a breaking change, but a pretty minor one since the `receiverhelper` metrics are all at `alpha` maturity, and since components using `otelcol_receiver_refused_xxx` to surface internal errors are arguably using the API "wrong"... Still, I think it may be good to err on the side of caution.

> What is this exactly? You mean to go with feature gate flag in the opentelemetry config right?

I'm talking about the `collector/featuregate` package, see https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate

jade-guiton-dd commented:
FYI, #13234 has now been merged, so I think the best way forward for this PR would be to use the new `consumererror.IsDownstream` wrapper instead of the custom one, and to place this new behavior under [a feature gate](https://github.com/open-telemetry/opentelemetry-collector/tree/main/featuregate) to avoid breaking users relying on the old semantics of `otelcol_receiver_refused_xxx`.

gizas commented on exporter/exporterhelper/internal/queue/persistent_queue_test.go:
```diff
@@ -58,6 +58,42 @@ func (int64Encoding) Unmarshal(bytes []byte) (context.Context, int64, error) {
 	return context.Background(), val, nil
 }
 
+func TestPersistentQueue_CorruptedItemOnRead(t *testing.T) {
```
Added for https://app.codecov.io/gh/open-telemetry/opentelemetry-collector/pull/12802/indirect-changes

jade-guiton-dd replied:
Hm, it seems pretty weird that your PR would affect the coverage of tests in the exporterhelper. Are you sure it's related to your changes, and isn't due to a change that was made on main?

If it's not related, I would rather we leave these tests out, and maybe file them as a separate PR so the people working on the exporterhelper can review them.

gizas replied:
Agree removed [moving and renaming featuregate and added otelcol_receiver_requests u…](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/89d7bc412c39d753f5b206a8b103931c66cba53e)

gizas commented on exporter/exporterhelper/xexporterhelper/profiles_test.go:
```diff
@@ -57,6 +57,20 @@ func TestProfilesRequest(t *testing.T) {
 	)
 }
 
+func TestProfilesEncoding_UnmarshalError(t *testing.T) {
```
Same for https://app.codecov.io/gh/open-telemetry/opentelemetry-collector/pull/12802/indirect-changes

gizas replied:
Removed [moving and renaming featuregate and added otelcol_receiver_requests u…](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/89d7bc412c39d753f5b206a8b103931c66cba53e)

gizas commented:
@jade-guiton-dd I think I have made some progress. I have made the choice to add a new flag additional to existing telemetry.newPipelineTelemetry
So if someone now wants to test this should run:

```bash
otelcol --config=config.yaml --feature-gates=+telemetry.newPipelineTelemetry,+telemetry.distinguishDownstreamErrors
```

<details><summary>Screenshot with some manual tests</summary>
I have even built and run some manual tests and I can see e2e the metrics appear:

<img width="2708" height="891" alt="Screenshot 2025-08-13 at 1 42 22 PM" src="https://github.com/user-attachments/assets/6646468b-6a02-4d65-9524-08d3c3c0b819" />
</details>

Let me know wdyt

gizas commented on receiver/receiverhelper/go.mod:
```diff
@@ -6,6 +6,9 @@ require (
 	github.com/stretchr/testify v1.10.0
 	go.opentelemetry.io/collector/component v1.38.0
 	go.opentelemetry.io/collector/component/componenttest v0.132.0
+	go.opentelemetry.io/collector/consumer/consumererror v0.0.0-00010101000000-000000000000
```
Done in [fixing version in go.mod](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/dd3df2384788d1b9d16cb8a21355c2b52516ba95)

gizas commented on internal/telemetry/telemetry.go:
```diff
@@ -22,6 +22,14 @@ var NewPipelineTelemetryGate = featuregate.GlobalRegistry().MustRegister(
+	featuregate.StageAlpha,
+	featuregate.WithRegisterFromVersion("v0.138.0"),
+	featuregate.WithRegisterDescription("Controls whether to distinguish downstream errors from internal errors, changing the 'outcome' for produced telemetry and splitting receiver metrics into 'refused' and 'failed' categories."),
+)
```
`receiverhelper.newReceiverMetrics` is the new one !

gizas commented:
@jade-guiton-dd I see now that the contrib-tests are failing. How do you want to proceed with this? 

<details><summary>Testing the failures</summary>

1. clone opentelemetry-collector-contrib in the same folder as opentelemetry-collector

```bash
cd opentelemetry-collector 
make prepare-contrib // This prepares the opentelemetry-collector-contrib to make use of the current checkout
make check-contrib // runs the tests
```

In general the idea is the same for all the tests that fail:
See for example this: https://github.com/open-telemetry/opentelemetry-collector-contrib/blob/main/receiver/awsxrayreceiver/receiver_test.go#L366

```yaml
got, err = tt.GetMetric("otelcol_receiver_failed_spans")
	assert.NoError(t, err)
	metricdatatest.AssertEqual(t,
		metricdata.Metrics{
			Name:        "otelcol_receiver_failed_spans",
			Description: "The number of spans that failed to be processed by the receiver due to internal errors.",
```

You should replace with above and all work and test passed
</details>

I have left the test running and can prepare an additional draft pr with all the related tests that fail. Any other suggestion?

gizas commented on receiver/receiverhelper/obsreport.go:
```diff
@@ -160,34 +161,57 @@ func (rec *ObsReport) endOp(
 	if err != nil {
 		numAccepted = 0
-		numRefused = numReceivedItems
+		if NewReceiverMetricsGate.IsEnabled() && consumererror.IsDownstream(err) {
```
Can you please make another check in [updating logic](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/b77091fb155c34169cdd22bc7b5c75b17e70c019)

I think I have updated it now once more but I want a fresh look. Will have another look tomorrow

gizas commented on receiver/receiverhelper/metadata.yaml:
```diff
@@ -28,17 +25,22 @@ telemetry:
+      unit: "{spans}"
+      sum:
+        value_type: int
+        monotonic: true
```
Done in [updating logic](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/b77091fb155c34169cdd22bc7b5c75b17e70c019)

jade-guiton-dd commented:
I'm surprised that test fails considering the feature gate should be disabled by default 🤔

gizas commented:
> Looking at the contrib-tests currently running in CI, I don't see any failures for receiver/awsxrayreceiver. I'm wondering if you weren't testing against an older version of your code, with the incorrect logic in obsreport?

Indeed @jade-guiton-dd our comments crossed. I wrote it the comment for the contrib-tests before seeing yours to update the logic. (I saw some errors today but unrelated to our changes). I have fixed the lint errors and retriggered the CI. Let's see now

gizas commented on cmd/otelcorecol/go.mod:
```diff
@@ -100,7 +100,7 @@ require (
 	go.opentelemetry.io/collector/connector/xconnector v0.132.0 // indirect
 	go.opentelemetry.io/collector/consumer v1.38.0 // indirect
-	go.opentelemetry.io/collector/consumer/consumererror v0.132.0 // indirect
+	go.opentelemetry.io/collector/consumer/consumererror v1.38.0 // indirect
```
Resolving this as I have fixed it in [fixing tests and lint errors](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/a1c0105c9e4c47ab3da6985fe2cfba8d10ab04a1)

gizas commented:
@jade-guiton-dd do you think are we ready to merge? anything else missing?

gizas commented on receiver/receiverhelper/metadata.yaml:
```diff
@@ -68,3 +79,29 @@ telemetry:
+      sum:
+        value_type: int
+        monotonic: true
+    receiver_requests:
```
Done in [adding alpha stability to receiver_requests](https://github.com/open-telemetry/opentelemetry-collector/pull/12802/commits/a8a0b2d1b4bc45f3766eb6663c0c94ce6eb08eaf)

> Thank you for your patience with all these back-and-forths.

Thank you so much for all this journey. For me was really a new knowledge ride
